BRMS tutorial: add `rstan` example (#346)
inbo/tutorials: content/tutorials/r_brms/brms_nl/workshop_1_mcmc_en_brms.Rmd

         ul_diff = elpd_diff  + qnorm(0.95) * se_diff)
```

### Conclusie

Zowel o.b.v. de PPC als vergelijkingen met verschillende model selectie criteria, kunnen we besluiten dat het tweede Poisson model met random intercepts het best past bij de data. In principe konden we dit ook verwachten op basis van onze eigen intuïtie en het design van de studie, nl. het gebruik van de Poisson distributie om aantallen te modelleren en het gebruik van random intercepts om te controleren voor een hiërarchisch design (habitats genest in sites).

## Deep Dive: `rstan`
### Stan: Wat? Waarom?!
De `brms` package is een gemaksverpakking van `rstan`, dat zelf de functionaliteit van `stan` naar R brengt.

> Stan is een modeleringsomgeving, geschreven in de programmeertaal `C`, die de benodigdheden probabilistische ("Bayesiaanse") modellen implementeerd.

Reviewer: ```suggestion
Stan is een modeleringsomgeving, geschreven in de programmeertaal `C`, die de benodigdheden voor probabilistische ("Bayesiaanse") modellen implementeerd.
```

Meer info vind je [op het internet](https://mc-stan.org).

> Het voordeel van `brms` is bruikbaarheid: veel functies werken "out-of-the-box", met redelijke standaardwaarden en een syntaxis die lijkt op wat frequentisten gewend zijn.

Reviewer: niet enkel frequentisten, het is de gangbare syntax voor veel packages in R (zowel frequentistische als bayesiaanse) die met statistische modellen werken

Author: Klopt! I generalized to "statisticians".

> Het relatieve gebruiksgemak gaat echter ten koste van de flexibiliteit en in zekere mate de leesbaarheid.

Reviewer: ```suggestion
Het relatieve gebruiksgemak kan echter ten koste gaan van de flexibiliteit en in zekere mate van de leesbaarheid.
```

Stan en `rstan` daarentegen leunen meer op de wiskundige formulering van modellen.

> Elk aspect van het model moet expliciet worden ingesteld, wat een voordeel kan zijn (bijvoorbeeld als u te maken hebt met niet-standaard use cases) of een nadeel (bijvoorbeeld als u modellen op niet-optionele manieren beveiligt).

Reviewer: ```suggestion
Elk aspect van het model moet expliciet worden ingesteld, wat een voordeel kan zijn (bijvoorbeeld als u te maken hebt met niet-standaard usecases) of een nadeel (bijvoorbeeld als u modellen op niet-optionele manieren beveiligt).
```

> Elk aspect van het model moet expliciet worden ingesteld, wat een voordeel kan zijn (bijvoorbeeld als u te maken hebt met niet-standaard use cases) of een nadeel (bijvoorbeeld als u modellen op niet-optionele manieren beveiligt).

Reviewer: wat bedoel je met "een model beveiligen"?

Author: This was a translation fault of a faulty English sentence, which on top of that I overlooked in proofreading. My apologies! It is corrected.

Om kort een indruk te geven, bouwen we dezelfde modellen als hierboven, met behulp van het Stan-framework.

```{r}
library("rstan")
conflicted::conflicts_prefer(rstan::extract)
conflicted::conflicts_prefer(brms::loo)
```

### Model Definition

> The simple poisson model resembles [one of the `stan`dard examples](https://mc-stan.org/docs/stan-users-guide/posterior-prediction.html#posterior-prediction-for-regressions), which you can refer to for all further details and more.

Reviewer: ```suggestion
Het eenvoudige Poisson-model lijkt op [een van de `stan`daardvoorbeelden](https://mc-stan.org/docs/stan-users-guide/posterior-prediction.html#posterior-prediction-for-regressions), waarnaar u kunt verwijzen voor alle verdere details.
```

Reviewer: je herhaalt de zin ook verderop, lijn 1084

Author: It was still in there from translating. thank you for the catch!

RMarkdown kan `stan`-chunks verwerken, hoewel de algemenere modeldefinitie is uitbesteed aan een apart "*.stan"-bestand.

> Alternatief kan je je model ook als grote tekstblok definiëren, zoals hieronder te zien.

Reviewer: ```suggestion
Alternatief kan je je model ook als een groot tekstblok definiëren, zoals hieronder te zien.
```

  vector[N] habitat_obs;            // habitat
  array[N] int<lower=0> sp_rich;    // outcome variable: number of ants
}

parameters {
  real intercept_rich;              // intercept
  real habitat_effect;              // slope
}

model {
  // priors
  intercept_rich ~ normal(0, 1);
  habitat_effect ~ normal(0, 1);

  // "posterior", "likelihood", ... give it a name!
  sp_rich ~ poisson_log(intercept_rich + habitat_effect * habitat_obs);

}

'
```

Zie dit als een "kijkje achter de schermen"!
Je moet expliciet de modelstructuur, priors en zelfs de gegevenstypen van invoervariabelen definiëren.
Maar let op hoe zelfs Stan niet zonder handige functies is: we kunnen de `poisson_log` posterior gebruiken om de $log(\lambda )$ te modelleren.

Als je buiten RStudio/RMarkdown werkt, kun je het model het beste laden vanuit een bestand:

```{r stan_load_model, eval=TRUE, class.source='fold-show'}
stan_poisson_model <- stan_model(
  # file = "./poisson_model.stan",
  model_code = stan_poisson_model_code, 
  model_name = "stan poisson model"
  )
```

### Sampling

Sampling doet vrijwel hetzelfde als hierboven, want in de kern is `brms` gewoon `stan`.

> ```{r stan_sampling}

Reviewer: Ik zou hier ook zelf een burn-in gebruiken zoals we in de brms-voorbeelden hebben gedaan. Dan is het nog meer gelijkaardig als hierboven.

Author: Thank you for pointing this out, @wlangera !
I briefly asked @hansvancalster  and googled "burn-in tuning warmup bayesian stats". There is a [to-the-point discussion here](https://statmodeling.stat.columbia.edu/2017/12/15/burn-vs-warm-iterative-simulation-algorithms). My notion of changing came from the fact that we don't physically "burn" or "burn in" anything; thus I was surprised that apparently the term was considered appropriate for some MCMC algorithms. "Tuning", my actual favorite :metal:, is reserved for a certain sub-part of the warm-up phase.

Anyways, the experts seem to prefer "warm-up", and if you agree and confirm, I will in turn adjust this throughout the tutorial.

Reviewer: Ok that sounds good. Thanks

stan_poisson_fit <- sampling(
  stan_poisson_model,
  list(
    N = nrow(ants_df),
    habitat_obs = as.integer(ants_df$habitat)-1,
    sp_rich = ants_df$sp_rich
    ),
  iter = niter,
  chains = nchains,
  cores = nparallel
  )

stan_poisson_fit
```

Een handige manier om modelfits snel te inspecteren is `shinystan`.
Het opent een browser met glanzende plots.

```{r eval=FALSE}
library("shinystan")
launch_shinystan(stan_poisson_fit)
```

En kijk: de modeluitkomst is precies zoals boven.
In het huidige geval is het voordeel van het overstappen op `stan` zeer beperkt.
In andere gevallen kan het lonen.

Weet dat Stan er voor u is, aarzel niet om de uitgebreide documentatie te raadplegen en wees niet bang om het te proberen!

### Huiswerk: Hierarchical Model
Om je modelleringsvaardigheden nog verder te ontwikkelen, kun je het "random intercept"-model implementeren en sampelen.

> In "Bayesiaanse" termen is het [algemene terminologie is "hiërarchisch" model](https://mc-stan.org/docs/stan-users-guide/regression.html#hierarchical-regression).

Reviewer: ```suggestion
In "Bayesiaanse" termen is de[algemene terminologie een "hiërarchisch" model](https://mc-stan.org/docs/stan-users-guide/regression.html#hierarchical-regression).
```
